metastore/src/java/org/apache/hadoop/hive/metastore/hbase/StatsCache.java

```diff
@@ -0,0 +1,324 @@
+
+/**
+ * Caching for stats.  This implements an LRU cache.   It does not remove entries explicitly as
+ * that is generally expensive to find all entries for a table or partition.  Instead it lets them
+ * time out. When the cache is full a sweep is done in the background to remove expired entries.
+ * This cache is shared across all threads, and so operations are protected by reader or writer
+ * locks as appropriate.
+ */
+class StatsCache  {
+  static final private Log LOG = LogFactory.getLog(StatsCache.class.getName());
+
+  private static StatsCache self = null;
+
+  private final long timeToLive;
+  private final int maxSize;
+  private Map<Key, StatsInfo> cache;
+  private ReadWriteLock lock;
+  private boolean cleaning;
+  private Counter tableMisses;
+  private Counter partMisses;
+  private Counter tableHits;
+  private Counter partHits;
+  private Counter cleans;
+  private List<Counter> counters;
+
+  static synchronized StatsCache getInstance(Configuration conf) {
+    if (self == null) {
+      int totalObjectsToCache =
+          HiveConf.getIntVar(conf, HiveConf.ConfVars.METASTORE_HBASE_CACHE_SIZE);
+      long timeToLive = HiveConf.getTimeVar(conf,
+          HiveConf.ConfVars.METASTORE_HBASE_CACHE_TIME_TO_LIVE, TimeUnit.SECONDS);
+      self = new StatsCache(totalObjectsToCache / 2, timeToLive);
+    }
+    return self;
+  }
+
+  /**
+   * @param max maximum number of objects to store in the cache.  When max is reached, eviction
+   *            policy is MRU.
+   * @param timeToLive time (in seconds) that an entry is valid.  After this time the record will
+   *                   discarded lazily
+   */
+  private StatsCache(int max, long timeToLive) {
+    maxSize = max;
+    this.timeToLive = timeToLive * 1000;
+    cache = new HashMap<Key, StatsInfo>();
+    lock = new ReentrantReadWriteLock();
+    cleaning = false;
+    counters = new ArrayList<Counter>();
+    tableMisses = new Counter("Stats cache table misses");
+    counters.add(tableMisses);
+    tableHits = new Counter("Stats cache table hits");
+    counters.add(tableHits);
+    partMisses = new Counter("Stats cache partition misses");
+    counters.add(partMisses);
+    partHits = new Counter("Stats cache partition hits");
+    counters.add(partHits);
+    cleans = new Counter("Stats cache cleans");
+    counters.add(cleans);
+  }
+
+  /**
+   * Add an object to the cache.
+   * @param dbName name of database table is in
+   * @param tableName name of table
+   * @param partName name of partition, can be null if these are table level statistics
+   * @param colName name of the column these statistics are for
+   * @param stats stats
+   * @param lastAnalyzed last time these stats were analyzed
+   */
+  void put(String dbName, String tableName, String partName, String colName,
+           ColumnStatisticsObj stats, long lastAnalyzed) {
+    if (cache.size() >= maxSize) clean();
+    Key key = new Key(dbName, tableName, partName, colName);
```

Comment:
should we avoid adding this key to cache if it is currently at the limit ? It is not certain that clean() call is going to free up any entries (all of them could return false for tooLate)